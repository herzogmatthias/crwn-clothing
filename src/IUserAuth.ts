export interface IUserAuth {
  id: string;
  displayName: string;
  email: string;
  createdAt: Date;
}
