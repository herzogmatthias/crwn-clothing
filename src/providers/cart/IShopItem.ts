export interface IShopItem {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
}
