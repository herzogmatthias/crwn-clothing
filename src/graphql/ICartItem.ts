import { IShopItem } from "./IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
