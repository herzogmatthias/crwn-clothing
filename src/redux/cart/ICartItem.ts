import { IShopItem } from "../shop/IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
