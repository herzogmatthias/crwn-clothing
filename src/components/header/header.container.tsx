import { Query } from "react-apollo";
import { GET_CART_HIDDEN } from "../../graphql/resolvers";
import { ApolloCurrentQueryResult } from "apollo-boost";
import React from "react";
import Header from "./header.component";
import { IHeaderType } from "../../graphql/types";

const HeaderContainer = () => {
  console.log("Hello");
  return (
    <Query query={GET_CART_HIDDEN}>
      {({ data, loading }: ApolloCurrentQueryResult<IHeaderType>) => {
        const { cartHidden } = data!;
        if (loading) return <Header hidden></Header>;
        return <Header hidden={cartHidden}></Header>;
      }}
    </Query>
  );
};

export default HeaderContainer;
