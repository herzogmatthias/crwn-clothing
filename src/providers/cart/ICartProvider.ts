import { ICartItem } from "../../redux/cart/ICartItem";
import { IShopItem } from "../../redux/shop/IShopItem";

export interface ICartProvider {
  hidden: boolean;
  total: number;
  toggleHidden(): void;
  cartItems: ICartItem[];
  addItem(item: IShopItem): void;
  removeItem(item: ICartItem): void;
  clearItemFromCart(item: ICartItem): void;
  cartItemsCount: number;
}
