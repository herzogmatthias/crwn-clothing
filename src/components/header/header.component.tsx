import * as React from "react";
import { ReactComponent as Logo } from "../../assets/crown.svg";
import { RootState } from "../../redux/root-reducer";
import { ConnectedProps, connect } from "react-redux";
import CartIcon from "../cart-icon/cart-icon.component";
import CartDropdown from "../cart-dropdown/cart-dropdown.component";
import { selectCurrentUser } from "../../redux/user/user.selectors";
import { selectCartHidden } from "../../redux/cart/cart.selectors";
import { createStructuredSelector } from "reselect";
import { IUserAuth } from "../../IUserAuth";
import {
  HeaderContainer,
  LogoContainer,
  OptionLink,
  OptionsContainer,
} from "./header.styles";
import { signOutStart } from "../../redux/user/user.actions";

interface ISelectorProps {
  currentUser: null | IUserAuth;
  hidden: boolean;
}

type IHeaderProps = ConnectedProps<typeof connector>;

export function Header({ currentUser, hidden, signOutStart }: IHeaderProps) {
  return (
    <HeaderContainer>
      <LogoContainer to="/">
        <Logo className="logo"></Logo>
      </LogoContainer>
      <OptionsContainer>
        <OptionLink to="/shop">SHOP</OptionLink>
        <OptionLink to="/contact">CONTACT</OptionLink>
        {currentUser ? (
          <OptionLink as="div" onClick={signOutStart}>
            SIGN OUT
          </OptionLink>
        ) : (
          <OptionLink to="/signIn">SIGN IN</OptionLink>
        )}
        <CartIcon></CartIcon>
      </OptionsContainer>
      {hidden ? null : <CartDropdown></CartDropdown>}
    </HeaderContainer>
  );
}
const mapDispatchToProps = {
  signOutStart: signOutStart,
};

const mapStateToProps = createStructuredSelector<RootState, ISelectorProps>({
  currentUser: selectCurrentUser,
  hidden: selectCartHidden,
});

const connector = connect(mapStateToProps, mapDispatchToProps);
export default connect(mapStateToProps, mapDispatchToProps)(Header);
