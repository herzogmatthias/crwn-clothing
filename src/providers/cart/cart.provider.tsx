import React, { createContext, useEffect } from "react";
import { ICartProvider } from "./ICartProvider";
import { ICartItem } from "../../redux/cart/ICartItem";
import {
  addItemToCart,
  removeItemFromCart,
  filterItemFromCart,
  getCartItemsCount,
  getTotalPrice,
} from "../../redux/cart/cart.utils";
import { IShopItem } from "../../redux/shop/IShopItem";

export const CartContext = createContext<ICartProvider>({
  hidden: true,
  total: 0,
  cartItems: [],
  cartItemsCount: 0,
  toggleHidden: () => {},
  addItem: () => {},
  removeItem: () => {},
  clearItemFromCart: () => {},
});

interface ICartProviderProps {
  children: React.ReactNode;
}

const CartProvider = ({ children }: ICartProviderProps) => {
  const [hidden, setHidden] = React.useState(true);

  const [cartItems, setCartItems] = React.useState<ICartItem[]>([]);
  const [cartItemsCount, setCartItemsCount] = React.useState(0);

  const [total, setTotal] = React.useState(0);

  const toggleHidden = () => {
    setHidden(!hidden);
  };

  const addItem = (item: IShopItem) =>
    setCartItems(addItemToCart(cartItems, item));

  const removeItem = (item: ICartItem) =>
    setCartItems(removeItemFromCart(cartItems, item));
  const clearItemFromCart = (item: ICartItem) =>
    setCartItems(filterItemFromCart(cartItems, item));

  useEffect(() => {
    setCartItemsCount(getCartItemsCount(cartItems));
    setTotal(getTotalPrice(cartItems));
  }, [cartItems]);

  return (
    <CartContext.Provider
      value={{
        total,
        addItem,
        hidden,
        toggleHidden,
        cartItems,
        cartItemsCount,
        removeItem,
        clearItemFromCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export default CartProvider;
