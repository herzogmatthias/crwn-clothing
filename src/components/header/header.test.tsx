import React from "react";
import { IUserAuth } from "../../IUserAuth";
import { default as ConnectedHeader, Header } from "./header.component";
import { ReactWrapper, mount, shallow, ShallowWrapper } from "enzyme";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { ICartItem } from "../../redux/cart/ICartItem";
import { ActionCreatorWithoutPayload } from "@reduxjs/toolkit";
import CartDropdown from "../cart-dropdown/cart-dropdown.component";
import { OptionLink } from "./header.styles";

describe("CONNECTED HEADER TESTS", () => {
  let wrapper: ReactWrapper;
  let mockStore = configureStore([]);
  let store: MockStoreEnhanced<unknown, {}>;
  const cartItems: ICartItem[] = [
    { id: 1, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
    { id: 2, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
  ];
  const currentUser: IUserAuth = {
    createdAt: new Date("22.05.2020"),
    id: "asdlfjas",
    displayName: "asdf",
    email: "asdf@gmail.com",
  };
  beforeEach(() => {
    const mockState = {
      user: {
        currentUser: currentUser,
      },
      cart: {
        hidden: true,
        cartItems: cartItems,
      },
    };

    store = mockStore(mockState);

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ConnectedHeader />
        </Provider>
      </BrowserRouter>
    );
  });
  test("render Directory Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("check for props", () => {
    expect(wrapper.find(Header).prop("hidden")).toEqual(true);
    expect(wrapper.find(Header).prop("currentUser")).toEqual(currentUser);
  });
});

describe("NON CONNECTED HEADER TESTS", () => {
  let wrapper: ShallowWrapper;
  const currentUser: IUserAuth = {
    createdAt: new Date("22.05.2020"),
    id: "asdlfjas",
    displayName: "asdf",
    email: "asdf@gmail.com",
  };
  let mockSignOutStart: ActionCreatorWithoutPayload<"SIGN_OUT_START">;
  beforeEach(() => {
    mockSignOutStart = jest.fn() as any;

    const mockProps = {
      hidden: true,
      currentUser: currentUser,
      signOutStart: mockSignOutStart,
    };

    wrapper = shallow(<Header {...mockProps} />);
  });

  test("render sign out link", () => {
    expect(wrapper.find(OptionLink).at(2).text()).toBe("SIGN OUT");
  });

  test("call signOutStart method when link is clicked", () => {
    wrapper.find(OptionLink).at(2).simulate("click");

    expect(mockSignOutStart).toHaveBeenCalled();
  });

  test("render sign in link if currentUser is null", () => {
    const mockProps = {
      hidden: true,
      currentUser: null,
      signOutStart: mockSignOutStart,
    };

    const newWrapper = shallow(<Header {...mockProps} />);

    expect(newWrapper.find(OptionLink).at(2).text()).toBe("SIGN IN");
  });

  test("don't render CartDropdown", () => {
    expect(wrapper.exists(CartDropdown)).toBe(false);
  });

  test("render CartDropdown if currentUser is null", () => {
    const mockProps = {
      hidden: false,
      currentUser: null,
      signOutStart: mockSignOutStart,
    };

    const newWrapper = shallow(<Header {...mockProps} />);

    expect(newWrapper.exists(CartDropdown)).toBe(true);
  });
});
