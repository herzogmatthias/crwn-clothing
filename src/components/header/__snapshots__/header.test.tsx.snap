// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONNECTED HEADER TESTS render Directory Component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(Header)>
        <Header
          currentUser={
            Object {
              "createdAt": Date { NaN },
              "displayName": "asdf",
              "email": "asdf@gmail.com",
              "id": "asdlfjas",
            }
          }
          hidden={true}
          signOutStart={[Function]}
        >
          <styled.div>
            <div
              className="sc-fzplWN kRkQAc"
            >
              <Styled(Link)
                to="/"
              >
                <Link
                  className="sc-fznyAO bVJWQd"
                  to="/"
                >
                  <LinkAnchor
                    className="sc-fznyAO bVJWQd"
                    href="/"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-fznyAO bVJWQd"
                      href="/"
                      onClick={[Function]}
                    >
                      <ForwardRef(SvgCrown)
                        className="logo"
                      >
                        <svg
                          className="logo"
                        >
                          crown.svg
                        </svg>
                      </ForwardRef(SvgCrown)>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
              <styled.div>
                <div
                  className="sc-fznKkj jFMUGs"
                >
                  <Styled(Link)
                    to="/shop"
                  >
                    <Link
                      className="sc-fznZeY cLBAIR"
                      to="/shop"
                    >
                      <LinkAnchor
                        className="sc-fznZeY cLBAIR"
                        href="/shop"
                        navigate={[Function]}
                      >
                        <a
                          className="sc-fznZeY cLBAIR"
                          href="/shop"
                          onClick={[Function]}
                        >
                          SHOP
                        </a>
                      </LinkAnchor>
                    </Link>
                  </Styled(Link)>
                  <Styled(Link)
                    to="/contact"
                  >
                    <Link
                      className="sc-fznZeY cLBAIR"
                      to="/contact"
                    >
                      <LinkAnchor
                        className="sc-fznZeY cLBAIR"
                        href="/contact"
                        navigate={[Function]}
                      >
                        <a
                          className="sc-fznZeY cLBAIR"
                          href="/contact"
                          onClick={[Function]}
                        >
                          CONTACT
                        </a>
                      </LinkAnchor>
                    </Link>
                  </Styled(Link)>
                  <Styled(Link)
                    as="div"
                    onClick={[Function]}
                  >
                    <div
                      className="sc-fznZeY cLBAIR"
                      onClick={[Function]}
                    >
                      SIGN OUT
                    </div>
                  </Styled(Link)>
                  <Connect(CartIcon)>
                    <CartIcon
                      itemCount={4}
                      toggleCartHidden={[Function]}
                    >
                      <styled.div
                        onClick={[Function]}
                      >
                        <div
                          className="sc-AxjAm fUffOI"
                          onClick={[Function]}
                        >
                          <Styled(Component)>
                            <ForwardRef(SvgShoppingBag)
                              className="sc-AxirZ iKeIkB"
                            >
                              <svg
                                className="sc-AxirZ iKeIkB"
                              >
                                shopping-bag.svg
                              </svg>
                            </ForwardRef(SvgShoppingBag)>
                          </Styled(Component)>
                          <styled.span>
                            <span
                              className="sc-AxiKw iNislX"
                            >
                              4
                            </span>
                          </styled.span>
                        </div>
                      </styled.div>
                    </CartIcon>
                  </Connect(CartIcon)>
                </div>
              </styled.div>
            </div>
          </styled.div>
        </Header>
      </Connect(Header)>
    </Provider>
  </Router>
</BrowserRouter>
`;
